### PR TITLE
Add links to static copies of vocabulary files and hashes

### PR DESCRIPTION
**Pull request by msporny**
This PR is an attempt at addressing issue #1261, which requested that cryptographic hashes be added to the vocabulary files. This PR utilizes both frozen static files and hash values to achieve the end result, which are stable vocabulary definitions for systems that perform RDF vocabulary processing (which is a completely optional, non-mandatory thing).

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-model/pull/1279.html" title="Last updated on Oct 21, 2023, 6:54 PM UTC (1a88c3c)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-model/1279/87a4db3...1a88c3c.html" title="Last updated on Oct 21, 2023, 6:54 PM UTC (1a88c3c)">Diff</a>

### CODE REVIEW

File: index.html
Comment (OR13) [inline index.html:5872]: this is good, but will obviously break the hash... the MAY here means we don't need to test it.

File: index.html
Comment (OR13) [inline index.html:5903]: does this mean that our use of `name` and `description` are normatively bound to version 22.0 of schema.org ? and if version 23.5 changes those term definitions verifiable credentials using the new definitions are not conformant?

File: index.html
Comment (OR13) [inline index.html:5915]: Is it possible to make this a W3C origin based URL?

File: index.html
Comment (OR13) [inline index.html:5927]: excellent

**Review by OR13 — CHANGES_REQUESTED** [review]
Requesting clarity regarding schema.org... Its possible that a few extra sentences specific to schema.org would address the issue.

File: index.html
Comment (msporny) [reply index.html:5872]: Yes, put that language in there to head off the inevitable: "But I want the vocabulary expressed in TURTLE or NQUADS" requests. This is the best compromise I could think of, open to other suggestions

File: index.html
Comment (msporny) [reply index.html:5903]: It means when you evaluate `name` and `description`, you are to use the v22.0 meaning of those properties.

This is the downside with using cryptographic hashes and specific releases, you don't end up using the latest meaning of the vocabulary. That said, I'd be surprised if `name` and `description` changed in any significant way in future releases.

File: index.html
Comment (msporny) [reply index.html:5915]: It will eventually be a date-stamped W3C URL. Since that URL doesn't exist today, this is all I have to work with.

The text at the top of the section highlights that this URL will be changing during CR.

File: index.html
Comment (OR13) [reply index.html:5903]: It seems like more of a downside that comes from using URLs for context files, we hope version 22.0 won't change terms, and that version 23.0 won't add terms we need, we have a way to detect if version 22 changes terms, and we have caching guidance from the remote server that applies to this.

I think we have done the best we can for URLs that reference context files, but if this thread motivates additional improvements mission accomplished.

Comment (OR13) [reply index.html:5903]: @msporny too close off this thread, do we need to add any updates regarding normative dependency on a specific version, or comment on caching headers? We had a thread somewhere else about how long to cache the w3c hosted v2 context, that is also relevant here.

File: index.html
Comment (OR13) [reply index.html:5915]: What will the eventual URL be?

https://www.w3.org/2023/data-integrity/vocabulary.jsonld ?

**Review by dlongley — APPROVED** [review]

File: index.html
Comment (OR13) [reply index.html:5915]: The only reason I raise it now, is we saw objections to github hosted documents in the DID WG, so if its possible to do a redirect soon, its one less thing that might come up later.

Comment (msporny) [reply index.html:5915]: The timing of this is largely up to @iherman and how quickly we get down the CR road.

File: index.html
Comment (BigBlueHat) [reply index.html:5869]: Here's the existing requirement from the [JSON Processing section](https://w3c.github.io/vc-data-model/#json-processing):
> Ensure that all values associated with a `@context` property are in the expected order, the contents of the context files match known good cryptographic hashes for each file, and domain experts have deemed that the contents are appropriate for the intended use case.

**Comment by BigBlueHat** [comment]
> Requesting clarity regarding schema.org... Its possible that a few extra sentences specific to schema.org would address the issue.

The versioned schema.org URL + hashing is equivalent to the versioned W3C URLs + hashing.

@OR13 are you wanting a statement that schema.org is in fact versioned, that there's a need to pick a specific published version for term stability, and that what is described in the table is doing exactly that?

If so, I can provide some suggested text--though it'll likely apply to any/all of the vocab URL (since schema.org isn't a unique animal). If not, I could use some additional thoughts on what you'd like to see clarified. Thanks!

**Review by Wind4Greg — APPROVED** [review]

File: index.html
Comment (davidlehn) [inline index.html:5916]: These don't seem to be correct right now.  If they are in flux, perhaps put in a temp FIXME value?

File: index.html
Comment (iherman) [reply index.html:5903]: My feeling is that we are overzealous at this point. We are using some absolute base terms of schema.org that is used at millions of places all over the world. The chances for changes are practically zero. In my view we can safely leave out schema.org from this whole consideration. After all, we also refer, indirectly, to xsd datatypes, rdf and rdfs core terms, and we do not put hash values for their respective URL-s either.

File: index.html
Comment (iherman) [reply index.html:5915]: Using github gives us the flexibility for easy changes. My preference would be to make a big cleanup on the URL-s when we are closing up to Proposed Rec, and not to complicate our life with this at this point.

**Review by Sakurann — CHANGES_REQUESTED** [review]
please remove schema.org parts as they are "part of fabric of RDF" (to borrow Ivan's words) and maybe clarify that the line of whether a URL is protected by hash or not is that and that the WG mainly focuses on protecting files defined by this WG (as opposed to those maintained by schema.org)

File: index.html
Comment (TallTed) [inline index.html:5872]:
```suggestion
serializations of the vocabulary files MAY be used by implementations.
```

File: index.html
Comment (TallTed) [inline index.html:5888]:
```suggestion
`application/ld+json`
```

File: index.html
Comment (TallTed) [inline index.html:5894]:
```suggestion
sha256: `z52TgKqh2nqTCuACI8lCvhRdjwxQjeVmuOMCDCEijq4=`<br><br>
sha3-512: `m8Ss+jgZiyL2Ws/ICJcWjHFd9PccJWsXPvMatBOhrH<wbr>
h0qCBrzfgO2zO1OQQbTL7zoPgLseIbcxJJpunD2bkoRA==`
```

File: index.html
Comment (TallTed) [inline index.html:5900]:
```suggestion
`application/ld+json`
```

File: index.html
Comment (TallTed) [inline index.html:5906]:
```suggestion
sha256: `KToiFier2I43VBuzvTkHG1dnBZEnpwoThC5AozLhnzA=`<br><br>
sha3-512: `uao4ahYBx+0pl5os+tg1CaljyXxVN4CpHtC19tESQF<wbr>
QYx7Rd5nQww/F8/ERpnsay1gDECReEyZzmffwhWV3omQ==`
```

File: index.html
Comment (TallTed) [inline index.html:5912]:
```suggestion
`application/ld+json`
```

File: index.html
Comment (TallTed) [inline index.html:5918]:
```suggestion
sha256: `LEaoTyf796eTaSlYWjfPe3Yb+poCW9TjWYTbFDmC0tc=`<br><br>
sha3-512: `f4DhJ3xhT8nT+GZ8UUZi4QC+HT//wXE2fRTgUP4UNw<wbr>
e4kvel2PFfd6jcofHBm9BjwEiGzVFGv4K+fFTKXRD2NA==`
```

**Review by TallTed — CHANGES_REQUESTED** [review]

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-09-14-vcwg#section1-9) on 2023-09-14

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 1.9. Add links to static copies of vocabulary files and hashes (pr vc-data-model#1279)

_See github pull request [vc-data-model#1279](https://github.com/w3c/vc-data-model/pull/1279)._




**Brent Zundel:** PR was raised 16 hours ago. Manu clearly is bored.  
� Can you give us an intro?  

**Manu Sporny:** We've made a decision to be more strict about how we specify the json-ld context and vocab documents used.  
� I believe it's because we want to be more strict about what a VC is.  
� We now are very specific for the semantics for all terms. And the contents of the files that describe it.  
� This PR does it in the vocabulary files themselves.  
� This PR points to the DI vocab and the VC vocab. This is us being very specific about which vocab documents are used.  

**Ivan Herman:** I'll just repeat here the same comment I made earlier.  
� If we start putting hash values in schema, then we'll have to put them everywhere. There is no chance that people will change existing items in some of these vocabularies. Getting to calculate the hash for a vocabulary is unnecessary.  

> *Kristina Yasuda:* +1 ivan.

**Ivan Herman:** It overcomplicated our lives to do so.  
� It's different for things we control. But for schema.org, we shouldn't do so.  

**Manu Sporny:** +1 to ivan. I wasn't a fan to use crypto hashes for vocab files. I wrote it because it was requested. Trying to be consistent. But where should we draw the line?  

> *Kristina Yasuda:* i am ok drawing the line whether this wg controls or not.

**Manu Sporny:** Definitely not for SSD, but let's not include schema.org from there.  
� I don't think it matters (?). I don't know if it matters that much.  
� Anyone using them can always decide to ignore the guidance.  
� I'm fine if we remove schema.org from the list, and only refer to vocab documents that this group is publishing.  
� Just noting that we are picking an arbitrary line.  

**Ivan Herman:** The way I would formulate it is that there are vocabs which are part of the core system/infrastructure. schema.org is one of them, along others. It's already part of the infrastructure.  

**Brent Zundel:** We have a path forward. Please others give feedback.  

</details>

File: index.html
Comment (msporny) [reply index.html:5869]: @BigBlueHat can you please provide a concrete change suggestion via "Add a suggestion" (CTRL-g)?

File: index.html
Comment (msporny) [reply index.html:5916]: They will be in flux for the forseeable future... I mentioned using FIXME values during one of the calls and it wasn't a popular option. Ideally, someone would write some code that retrieves these files and does the hash within the spec, but until that happens, this is going to be a manual process and perpetually broken until we get into CR.

**Comment by msporny** [comment]
@Sakurann @OR13 and @iherman -- I'm struggling to reconcile your requests. @iherman feels that we're being too overzealous with providing a hash for schema.org. @Sakurann would like us to remove it entirely, I think? @OR13 wants us to elaborate more on the hash associated with schema.org and speak to how long to cache it and/or how changes might be made.

I'm ambivalent to what we do here as I don't think that adding hashes for the vocabulary files is that useful. Splitting the difference between what each of you have asked for, would this work for all of you:

1. Remove the hash for schema.org. (to address @Sakurann and @iherman's request)
2. Say why we don't include the hash for schema.org (it changes on a regular basis and is viewed as a fairly stable vocabulary with good change control). (to address @OR13's request)

Thoughts?

**Comment by iherman** [comment]
> 1. Remove the hash for schema.org. (to address @Sakurann and @iherman's request)
> 2. Say why we don't include the hash for schema.org (it changes on a regular basis and is viewed as a fairly stable vocabulary with good change control). (to address @OR13's request)
> 
> Thoughts?

One more point to schema.org. It is a very large and, most importantly, ever-increasing vocabulary, but it has a very stable "core". Adding a hash to the _full_ schema.org vocabulary is pointless, and would become obsolete very quickly. So far, we make use of the stable core, though, which has been around in the vocabulary more or less from the start. Those terms are, for our purposes, cast in concrete. I just do not see any hash would ever bring anything important.

File: index.html
Comment (BigBlueHat) [inline index.html]:
```suggestion
Implementations that depend on RDF vocabulary processing MUST that all values associated
with a `@context` property are in the expected order, the contents of the context files match
known good cryptographic hashes for each file, and domain experts have deemed that the
contents are appropriate for the intended use case. Other semantically equivalent serializations
of the vocabulary files MAY be utilized by implementations.
```

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-10-11-vcwg#section1-12) on 2023-10-11

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 1.12. Add links to static copies of vocabulary files and hashes (pr vc-data-model#1279)

_See github pull request [vc-data-model#1279](https://github.com/w3c/vc-data-model/pull/1279)._




**Brent Zundel:** Raised by Manu.  

**Manu Sporny:** I don't know what to do with this PR just yet.  
… It felt like Kristina, Orie, and Ivan were asking for some irreconcilable things, I imagine we need the three of them on the call to figure out what to do here. I think Kristina and Ivan might be agreeing to remove the schema.org stuff -- I can't tell if Orie would be ok.  
… I think that's where we are ... some concrete text from Orie could resolve this or time on a call to discuss.  

**Brent Zundel:** Thank you, Manu.  
… We're at the point of trying to get to CR, with a PR like this, it's been raised, people have come back and say "I don't like it, I want changes" ... we're running out of time for back and forth. We don't have time for requests for changes, review of that, responses to that, responses to those responses, etc.  
… This group will need to decide soon on a shifting of our work mode to move through these things. If a PR doesn't gain consensus quickly we will have to look at just closing it if we don't have the capacity to address it.  
… Having soapboxed for a moment there, let's jump into 1276.  

</details>

**Comment by msporny** [comment]
Path forward discussed on the call today:

1. Remove hashes for schema.org. (to address @iherman and @Sakurann)
2. Add text noting that schema.org changes on a regular basis and is viewed as a fairly stable vocabulary with good change control. (to address @OR13)

I will implement this in the PR.

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-10-17-vcwg-special#section2-4) on 2023-10-17

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 2.4. Add links to static copies of vocabulary files and hashes (pr vc-data-model#1279)

_See github pull request [vc-data-model#1279](https://github.com/w3c/vc-data-model/pull/1279)._




**Brent Zundel:** This one is possibly on a path toward consensus?  
… Orie wanted clarity around schema.org ... Kristina is requesting changes. Kristina wants removal of the schema.org parts, Ivan also has some concerns.  

**Manu Sporny:** Ivan, it's not clear to me what you want done for this PR. It seemed like you were out of sync with what Kristina and Orie were asking for.  
… I think what's being requested is to not add hashes for schema.org.  

**Ivan Herman:** That's what I want as well.  

**Manu Sporny:** The clear path forward here is to not provide hashes for schema.org and Orie wanted some text about why.  

**Brent Zundel:** I think the explanatory line you had in your comment would be fine.  
… schema.org changes on a regular basis but uses change control and is stable, etc. therefore we do not provide a hash.  

**Ivan Herman:** The same thing we do with dublin core terms would apply for schema.org.  
… Then we are fine there.  
… If someone comes in with a term that is a dublin core term, which is fine, the same would apply as for schema.org.  

**Brent Zundel:** Thank you, Ivan.  
… Any other comments?  
… We believe we have consensus but, within a week, again ... if others come along and say we don't have consensus, we will close it.  

</details>

**Review by iherman — APPROVED** [review]

**Review by TallTed — CHANGES_REQUESTED** [review]
Small editorial tweaks to the note.

**Comment by msporny** [comment]
Still waiting on a re-review from @Sakurann and @OR13.

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-10-31-vcwg-special#section1-10) on 2023-10-31

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 1.10. Add links to static copies of vocabulary files and hashes (pr vc-data-model#1279)

_See github pull request [vc-data-model#1279](https://github.com/w3c/vc-data-model/pull/1279)._




**Brent Zundel:** Add links to static copies of vocab files and hashes.  
… This is another one of those that -- I thought we had a path forward.  
… Changes have been requested, changes have been made, we haven't had a re-review from the requesters.  

**Manu Sporny:** Yes, we're waiting on Kristina and Orie to re-review and approve.  

**Brent Zundel:** But we've been waiting for re-review for two weeks?  

**Manu Sporny:** You are correct.  

**Brent Zundel:** Ok, then it's entirely appropriate for us to say that we've done everything we can to address those requests, in the absence of re-review we can only assume we've done things correctly and we will merge today.  

**Ivan Herman:** I believe that the changes you've changes will address the comment that was more or less the same between Kristina and myself. She didn't +1 but you answered her comment, that's how I see it. I cannot comment on Orie's requests I didn't look at the details.  

**Brent Zundel:** Once the minutes are in I plan on merging this PR and the minutes will note that.  
… Any comments before moving on?  

</details>

**Comment by brentzundel** [comment]
As discussed in the 31 October meeting, this PR has been reviewed and requested changes have been made in good faith. Therefore I am merging it  as opportunity for re-review has been available for two weeks.

**Pull request merged**
